### PR TITLE
Controlling unfolding and evaluation order in the normalizer based on a new `strict_on_arguments` attribute

### PR DESCRIPTION
**Pull request by nikswamy**
We often mark certain definitions for unfolding during normalization but currently can only do so conditionally. 

In issue #1674, I proposed a new construct `norm_if_guard`  for conditional unfolding, but enthusiasm for that level of control seems low. Perhaps it's still worth discussing that there.

Meanwhile, this pull request provides support for a limited form of conditional unfolding based on tagging a definition with an attribute.

From FStar.Pervasives.fst:

```
(**
 * This attribute is used to control the evaluation order
 * and unfolding strategy for certain definitions.
 *
 * In particular, given 
 *    [@(strict_on_arguments [1;2])]
 *    let f x0 (x1:list x0) (x1:option x0) = e
 *
 * An application `f e0 e1 e2` is reduced by the normalizer by:
 *   1. evaluating e0 ~>* v0, e1 ~>* v1, e2 ~>* v2
 *
 *   2 a.
 *      If, according to the positional arguments [1;2], 
 *      if v1 and v2 have constant head symbols 
 *             (e.g., v1 = Cons _ _ _, and v2 = None _)
 *      then `f` is unfolded to `e` and reduced as
 *        e[v0/x0][v1/x1][v2/x2]
 *       
 *   2 b. 
 *
 *     Otherwise, `f` is not unfolded and the term is `f e0 e1 e2`
 *     reduces to `f v0 v1 v2`.
 *
 *)
irreducible
let strict_on_arguments (x:list int) : unit = ()
```

For examples of this attribute in use see FStar.Integers.fst and examples/micro-benchmarks./StrictUnfolding.fst. I'm reproducing some of that here.

You can mark this projector with an attribute to unfold it only when applied to term that will actually cause the match in its body to reduce.
 
```
[@(strict_on_arguments [1])]
let project #a (x:option a{Some? x}) =
  match x with
  | Some v -> v
```

Now, here are some tests to see how it works:

```
let test1 (x:option (x:int{x = 0}){Some? x}) =
  assert (project x == 0) //this doesn't reduce since x does not have a constant head symbol
    by (dump "A";
        norm [delta; iota];
        dump "B";
        smt ()) //so the proof has to go to SMT, which can, of course reason about the definition
```

But, 

```
let test2 () =
  assert (project (Some 0) == 0) //this one reduces to 0, since the head symbol is `Some`
    by (dump "A";
        norm [delta; iota];
        dump "B";
        trefl()) //and is dispatched by reflexivity
```
 
A big part of the motivation for adding this feature was because of @karthikbhargavan observing that integer-generic code extracts poorly after inlining. So, part of this pull request is a revision of `FStar.Integers` decorating its definitions with this new attribute so that the code doesn't blow up in size when extracting. 

For example, we now have:

```
module FStar.Integers
... 
[@mark_for_norm (strict_on_arguments [0])]
unfold
let ( + ) #sw
          (x:int_t sw)
          (y:int_t sw{within_bounds sw (v x + v y)})
  : Tot   (z:int_t sw)
``` 

Here are couple of examples illustrating it at work:

```

(* the code below extracts with `fstar --codegen OCaml` to
let (test_extraction_generic :
  FStar_Integers.signed_width -> Obj.t -> Obj.t -> Obj.t -> Obj.t -> Obj.t) =
  fun sw  ->
    fun w  ->
      fun x  ->
        fun y  ->
          fun z  ->
            FStar_Integers.op_Plus sw w
              (FStar_Integers.op_Plus sw x (FStar_Integers.op_Plus sw y z))
*)
let test_extraction_generic #sw (w x y z:int_t sw) : int_t sw =
  assume (ok (+) y z);
  assume (ok (+) x (y + z));
  assume (ok (+) w (x + (y + z)));
  w + (x + (y + z))

(* the code below extracts with `fstar --codegen OCaml` to

let (test_extraction_specific :
  FStar_UInt64.t ->
    FStar_UInt64.t -> FStar_UInt64.t -> FStar_UInt64.t -> FStar_UInt64.t)
  =
  fun w  ->
    fun x  ->
      fun y  ->
        fun z  ->
          FStar_UInt64.add w (FStar_UInt64.add x (FStar_UInt64.add y z))
*)
let test_extraction_specific (w x y z:uint_64) : uint_64 =
  assume (ok (+) y z);
  assume (ok (+) x (y + z));
  assume (ok (+) w (x + (y + z)));
  w + (x + (y + z))
```

@karthikbhargavan @s-zanella : do you want to try this out in Lib.IntTypes?

### CODE REVIEW

**Comment by s-zanella** [comment]
Thanks, this looks exactly what we need. I'll try it on `Lib.IntTypes`

**Comment by s-zanella** [comment]
One question. This doesn't work well with interfaces (or `inline_for_extraction`), does it? 
To get the desired behavior one would need to add an auxiliary definition in the implementation marked as `unfold`:
- .fsti
```FStar
val f (x y) : int
```
- .fst:
```FStar
[@(strict_on_arguments [0])]
unfold
let _f x y = ...

let f x y = _f x y
```

File: src/typechecker/FStar.TypeChecker.Normalize.fs [hunk 1]
```diff
@@ -1214,9 +1214,50 @@ let rec norm : cfg -> env -> stack -> term -> term =
...
+              norm cfg env stack head
+
+            | Some strict_args ->
+              // BU.print2 "%s has strict args [%s]\n"
+              //   (Print.term_to_string head)
+              //   (List.map string_of_int strict_args |> String.concat "; ");
+              let norm_args = args |> List.map (fun (a, i) -> (norm cfg env [] a, i)) in
+              if strict_args
+                |> List.for_all (fun i ->
+                  let arg_i, _ = List.nth norm_args i in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.Normalize.fs]: Should we check here that `List.length norm_args > i` (which could happen in case it is partially applied or programmer makes a mistake?).

File: src/smtencoding/FStar.SMTEncoding.EncodeTerm.fs [hunk 2]
```diff
@@ -101,6 +101,13 @@ let whnf env t =
 let norm env t = N.normalize [Env.Beta; Env.Exclude Env.Zeta;  //we don't know if it will terminate, so no recursion
                               Env.Eager_unfolding; Env.EraseUniverses] env.tcenv t
 
+let maybe_whnf env t =
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.EncodeTerm.fs]: I understood the other changes, but don't really get why smt encoding needs to change. Can you please add some comments to the code on how are these related to the strict arguments requirement?

**Comment by aseemr** [comment]
Left some minor comments, looks good to me otherwise, thanks!

**Comment by s-zanella** [comment]
I tried it on `Lib.IntTypes` and it does solve the code size issue. The pattern that worked for me is **not** marking the `val` declaration as `inline_for_extraction` in the interface, and introducing an auxiliary definition marked as `unfold` and strict. E.g.
- Lib.IntTypes.fsti
```FStar
val logor: #t:inttype -> #l:secrecy_level-> int_t t l -> int_t t l-> int_t t l
```
- Lib.IntTypes.fst
```FStar
[@(strict_on_arguments [0])]
unfold
val _logor: #t:inttype -> #l:secrecy_level -> int_t t l -> int_t t l -> int_t t l
let _logor #t #l a b = ...

let logor = _logor
```

File: src/typechecker/FStar.TypeChecker.Normalize.fs (hunk 1 above)
Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.Normalize.fs]: oh, oops. good catch, thanks!

**Comment by nikswamy** [comment]
One of the quirks of this feature is that it doesn't yet deal with partial applications well. 

E.g., if you have this:

```
[@(strict_on_arguments [1])]
let project a (x:option a{Some? x}) =
  match x with
  | Some v -> v
```

And you write `normalize_term project`,
then this will get unfolding to `fun a x -> match x with | Some v -> x`

I am considering enforcing in the typechecker that if a definition is marked `strict_on_arguments [..n...]`then it cannot be applied to fewer than `n` arguments.

WDYT? @aseemr @s-zanella et al?

**Comment by s-zanella** [comment]
> I am considering enforcing in the typechecker that if a definition is marked strict_on_arguments [..n...]then it cannot be applied to fewer than n arguments.

I think this is a reasonable restriction.

What about `inline_for_extraction`. Do you have a simpler pattern that works with interfaces or do we have to introduce auxiliary definitions and mark those as strict?

**Comment by nikswamy** [comment]
Oh, for the record, here's a repro of the problem with interfaces and inline for extraction.

Interface:

```
module StrictUnfolding.Two
[@(strict_on_arguments [1])]
inline_for_extraction
val project (#a:Type) (x:option a{Some? x}) : Tot a
```

Implementation:
```
module StrictUnfolding.Two
let project #a x = match x with | Some v -> v
```

Client

```
module StrictUnfolding.Client
open StrictUnfolding.Two
let test () : int = project (Some 0)
let test2 (x:option int{Some? x}) : int = project x
```

And

```
$ fstar --cmi StrictUnfolding.Client.fst --codegen OCaml
Extracted module FStar.Pervasives.Native
Extracted module FStar.Pervasives
Extracted module StrictUnfolding.Two
Extracted module StrictUnfolding.Client
Verified module: StrictUnfolding.Client (0 milliseconds)
All verification conditions discharged successfully

$ cat StrictUnfolding_Client.ml
open Prims
let (test : unit -> Prims.int) = fun uu____6  -> (Prims.parse_int "0")
let (test2 : Prims.int FStar_Pervasives_Native.option -> Prims.int) =
  fun x  -> match x with | FStar_Pervasives_Native.Some v -> v
```

My guess is that the attribute is being dropped in the interface file somehow. Looking into it

**Comment by nikswamy** [comment]
Hi @s-zanella. This works now:

```
$ cat StrictUnfolding.Two.fsti
module StrictUnfolding.Two

[@(strict_on_arguments [1])]
inline_for_extraction
val project (#a:Type) (x:option a{Some? x}) : Tot a

$ cat StrictUnfolding.Two.fst
module StrictUnfolding.Two

let project #a x = match x with | Some v -> v

$ cat StrictUnfolding.Client.fst
module StrictUnfolding.Client
open StrictUnfolding.Two

let test1 (x:option int{Some? x}) : int = project x
let test2 () : int = project (Some 0)

$ cat StrictUnfolding_Client.ml
open Prims
let (test1 : Prims.int FStar_Pervasives_Native.option -> Prims.int) =
  fun x  -> StrictUnfolding_Two.project x
let (test2 : unit -> Prims.int) = fun uu____21  -> (Prims.parse_int "0")
```

**Comment by s-zanella** [comment]
There still seems to be a problem with attributes not being copied from `val` to `let` in the implementation of the same module, rather than on clients of the module.

File: src/smtencoding/FStar.SMTEncoding.EncodeTerm.fs (hunk 2 above)
Comment (nikswamy) [reply src/smtencoding/FStar.SMTEncoding.EncodeTerm.fs]: Added this comment:

```
(* `maybe_whnf env t` attempts to reduce t to weak-head normal form.
 *  It is called when `t` is a head redex, e.g., if its head symbol is marked for unfolding.
 *  However, if its head symbol is also marked as `strict_on_arguments`, then if it is applied
 *  to non-constant arguments, then it may actually not be unfolded.
 *  In those cases `maybe_whnf env t` may not reduce `t` at all.
 *  In callers of this code, we need to be careful to check that if `t` was not reduced, then
 *  we do not enter into infinite loops by recursing on `t` itself.
 *)
```

File: src/typechecker/FStar.TypeChecker.Normalize.fs (hunk 1 above)
Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.Normalize.fs]: I added a bounds check. thanks

**Comment by nikswamy** [comment]
I have a local everest green and am going to merge this PR before things diverge again.

That said, there are still a couple of open issues left on the PR

1. I am not yet enforcing that symbols are applied to at least the arity mentioned in their `strict_on_arguments` attribute (if any). I plan to add this soon, though.

2. I have yet to reproduce @s-zanella's issue with attributes not being propagated within the val/let of a module. Since he seems to have a workaround, I'd prefer to merge this now and then revisit the issue

**Comment by s-zanella** [comment]
Please merge this as is, none of the issues are blocking.

Here's a modification of your example that shows the issue with propagating `strict_on_arguments` within an implementation of a module:

- `M.fsti`
```FStar
module M

[@(strict_on_arguments [1])]
inline_for_extraction
val project (#a:Type) (x:option a{Some? x}) : Tot a
```
- `M.fst`
```FStar
module M

let project #a x = match x with | Some v -> v

let test1 (x:option int{Some? x}) : int = project x
```
```FStar
module N

let test1 (x:option int{Some? x}) : int = M.project x
```
- `N.ml`
```OCaml
open Prims
let (test1 : Prims.int FStar_Pervasives_Native.option -> Prims.int) =
  fun x -> M.project x
```
- `M.ml`
```OCaml
open Prims
let project : 'Aa . 'Aa FStar_Pervasives_Native.option -> 'Aa =
  fun x -> match x with | FStar_Pervasives_Native.Some v -> v
let (test1 : Prims.int FStar_Pervasives_Native.option -> Prims.int) =
  fun x -> match x with | FStar_Pervasives_Native.Some v -> v
```
